### DISCUSSION TITLE
Carrying context across requests.

### DISCUSSION THREAD

**Discussion by shivchelwa** [open-telemetry/opentelemetry-java#3174] [Q&A]
I am tracing http requests on my server application. As this request is received by server, span is created for each component the request traverses. The traces are visible in Jaeger and request can be traced just fine. 

Now there is a requirement where I have to trace progress of an order across multiple http requests. Along with the http request tracing(which create and update order state), I create a span to trace the order state progress. On the very first state, I use root context to create a span(so as to use a different traceid than current context). And as the order state changes on subsequent  http request(may be different os process altogether in a clustered setup), I store the context on the order object using propagator. On the next request as the state changes, context is extracted from the order object, previous span is ended and new span is created using this context. Thus, multiple spans are created in this pattern and as the order is fulfilled, the last span is ended. I do not see any errors but these spans are not exported as a search by trace-id does not yield any results.

Am I missing anything? Is there any issue with usage of API?

**Comment by jkwatson** [comment]
I'd have to see some code and try it out to see what might be going wrong. This is definitely an unusual usage of the API, but I don't see anything in principle that wouldn't work. Can you enable the logging exporter with a SimpleSpanProcessor to make sure the spans are actually getting ended properly?

**Reply by shivchelwa** [reply]
@jkwatson, here is some sample code.
```
// Sample code when a request is received to change order state to given state(say CREATED)
// Create a new span
Context context = txtPropagator.extract(TelemetryHelper.rootContext(), custOrder, ORDER_CONTEXT_GETTER);
Span span = getTracer()
    .spanBuilder(spanName(custOrder))
    .setParent(context)
    .setSpanKind(spanKind())
    .startSpan();
context = context.with(span);
txtPropagator.inject(context, custOrder, ORDER_CONTEXT_SETTER);


// Sample code when a request is received to change order state from CREATED to SCHEDULED(above code is called, right after ending span)
// End the span
Context context = txtPropagator
    .extract(TelemetryHelper.rootContext(), custOrder, ORDER_CONTEXT_GETTER);
Span span = Span.fromContext(context);
span.end();
```

As I mentioned the span is ended when the state of an order changes on next request. So, I retrieve the span from an extracted context and I see I get a PropagatedSpan instead of RecordEventsReadableSpan. When I checked source, nothing happens on PropagatedSpan.end(). How do I handle this?

**Reply by jkwatson** [reply]
A Propagated Span isn't one that would ever get exported. Its purpose is purely just to act as a parent for other spans, as it has been created externally to the process. If you're pulling a span off the wire, it's one that was created by another system, and not the responsibility of the SDK to export. Does that make sense?

In general, propagators are not in the business of creating actual spans. They are just there to pull an external span off the wire, and set it into the Context for parenting purposes. (and of course, to do the reverse on outgoing calls).

**Reply by shivchelwa** [reply]
> If you're pulling a span off the wire, it's one that was created by another system, and not the responsibility of the SDK to export.

Can you please elaborate. Does that mean, span has to be created and ended by same process? Can it not be recreated in another system and then end it?

**Reply by jkwatson** [reply]
That is true. You definitely cannot end a span that was created in another system/process.
